File: src/AutomatonTheory/JavaLogic/FileManager.java
package AutomatonTheory.JavaLogic;

public class FileManager {//implements IFileManager// {
//
//    String ln = System.lineSeparator();
//    @Override
//    public Automaton loadAutomaton(String automatonName) {
//        Automaton returnAutomaton = null;
//
//        String allText = "";
//        String line;
//        try (
//                InputStream fis = new FileInputStream(automatonName + ".txt");
//                InputStreamReader isr = new InputStreamReader(fis, Charset.forName("UTF-8"));
//                BufferedReader br = new BufferedReader(isr);
//        ) {
//            while ((line = br.readLine()) != null) {
//                if(line.contains("Type")){
//                    String[] elements = line.split(":");
//                    if(elements[1].equals("DFA")){
//                        returnAutomaton = new DeterministicFiniteAutomaton(automatonName);
//                    }
//                    if(elements[1].equals("NFA")){
//                        returnAutomaton = new NonDeterministicFiniteAutomaton(automatonName);
//                    }
//                    if(elements[1].equals("NFAe")){
//                        returnAutomaton = new NonDeterministicFiniteEpsilonAutomaton(automatonName);
//                    }
//                }
//                if(line.contains("Alphabet")){
//                    String[] elements = line.split(":");
//                    String[] symbols = elements[1].split(";");
//                    for(int i = 0; i < symbols.length ; i++){
//                        returnAutomaton.Alphabet.add(symbols[i]);
//                    }
//                }
//                if(line.contains("States")){
//                    String[] elements = line.split(":");
//                    String[] states = elements[1].split(";");
//                    for(int i = 0; i < states.length ; i++){
//                        String[] attributes = states[i].split(",");
//                        returnAutomaton.States.add(new State(attributes[0], (attributes[1].equals("true"))?true:false, (attributes[2].equals("true"))?true:false));
//                    }
//                }
//                if(line.contains("Transitions")){
//                    String[] elements = line.split(":");
//                    String[] transitions = elements[1].split(";");
//                    for(int i = 0; i< transitions.length ; i++){
//                        String [] attributes = transitions[i].split(",");
//                        State currentState = returnAutomaton.getState(attributes[0]);
//                        currentState.addTransition(new Transition(returnAutomaton.getState(attributes[1]),attributes[2]));
//                    }
//                }
//
//            }
//        } catch (FileNotFoundException e) {
//            e.printStackTrace();
//        } catch (IOException e) {
//            e.printStackTrace();
//        }
//
//        return returnAutomaton;
//    }
//
//    @Override
//    public boolean saveAutomaton(Automaton automaton) {
//        try{
//            FileWriter fw = new FileWriter(automaton.AutomatonName + ".txt");
//
//            fw.write("Type:" + automaton.Type.toString() + ln);
//            fw.write("Alphabet:");
//            for(int i = 0; i < automaton.Alphabet.size() ; i++){
//                fw.write(automaton.Alphabet.get(i));
//                if(i < automaton.Alphabet.size()-1){
//                    fw.write(";");
//                }
//            }
//            fw.write(ln);
//
//            fw.write("States:");
//            for(int i = 0; i < automaton.States.size() ; i++){
//                State state = automaton.States.get(i);
//                fw.write(state.Name + "," + state.InitialState + "," + state.AcceptanceState);
//                if(i < automaton.Alphabet.size()-1){
//                    fw.write(";");
//                }
//            }
//            fw.write(ln);
//
//            fw.write("Transitions:");
//
//            for(int i = 0; i < automaton.States.size() ; i++){
//                State state = automaton.States.get(i);
//                for( int j=0 ; j < state.Transitions.size() ; j++){
//                    Transition transition = state.Transitions.get(j);
//                    if(automaton.States.size()-1 == i && state.Transitions.size()-1 == j){
//                        fw.write(state.Name + "," + transition.DestinyState.Name + "," + transition.Symbol);
//                    }
//                    else{
//                        fw.write(state.Name + "," + transition.DestinyState.Name + "," + transition.Symbol + ";");
//                    }
//
//                }
//            }
//
//            fw.close();
//        }
//        catch(Exception e){
//            e.printStackTrace();
//        }
//        return false;
//    }
}

File: src/AutomatonTheory/Kotlin/FileController/FileManager.kt
class FileManager {//: IFileManager {

//    internal var ln = System.lineSeparator()
//    @Throws(IOException::class)
//    override fun loadAutomaton(automatonName: String): Automaton {
//        var returnAutomaton: Automaton? = null
//
//        val allText = ""
//        var line: String
//        try {
//            FileInputStream(automatonName + ".txt").use { fis ->
//                InputStreamReader(fis, Charset.forName("UTF-8")).use { isr ->
//                    BufferedReader(isr).use { br ->
//                        line = br.readLine()
//                        while ((line = br.readLine()) != null) {
//                            if (line.contains("Type")) {
//                                val elements = line.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                if (elements[1] == "DFA") {
//                                    returnAutomaton = DeterministicFiniteAutomaton(automatonName)
//                                }
//                                if (elements[1] == "NFA") {
//                                    returnAutomaton = NonDeterministicFiniteAutomaton(automatonName)
//                                }
//                                if (elements[1] == "NFAe") {
//                                    returnAutomaton = NonDeterministicFiniteEpsilonAutomaton(automatonName)
//                                }
//                            }
//                            if (line.contains("Alphabet")) {
//                                val elements = line.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                val symbols = elements[1].split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                for (i in symbols.indices) {
//                                    returnAutomaton!!.Alphabet.add(symbols[i])
//                                }
//                            }
//                            if (line.contains("States")) {
//                                val elements = line.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                val states = elements[1].split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                for (i in states.indices) {
//                                    val attributes = states[i].split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                    returnAutomaton!!.States.add(State(attributes[0], if (attributes[1] == "true") true else false, if (attributes[2] == "true") true else false))
//                                }
//                            }
//                            if (line.contains("Transitions")) {
//                                val elements = line.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                val transitions = elements[1].split(";".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                for (i in transitions.indices) {
//                                    val attributes = transitions[i].split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
//                                    val currentState = returnAutomaton!!.getState(attributes[0])
//                                    currentState.addTransition(Transition(returnAutomaton!!.getState(attributes[1]), attributes[2]))
//                                }
//                            }
//
//                        }
//                    }
//                }
//            }
//        } catch (e: FileNotFoundException) {
//            e.printStackTrace()
//        } catch (e: IOException) {
//            e.printStackTrace()
//        }
//
//        return returnAutomaton
//    }
//
//    override fun saveAutomaton(automaton: Automaton): Boolean {
//        try {
//            val fw = FileWriter(automaton.AutomatonName + ".txt")
//
//            fw.write("Type:" + automaton.Type.toString() + ln)
//            fw.write("Alphabet:")
//            for (i in 0..automaton.Alphabet.size - 1) {
//                fw.write(automaton.Alphabet[i])
//                if (i < automaton.Alphabet.size - 1) {
//                    fw.write(";")
//                }
//            }
//            fw.write(ln)
//
//            fw.write("States:")
//            for (i in 0..automaton.States.size - 1) {
//                val state = automaton.States[i]
//                fw.write(state.Name + "," + state.InitialState + "," + state.AcceptanceState)
//                if (i < automaton.Alphabet.size - 1) {
//                    fw.write(";")
//                }
//            }
//            fw.write(ln)
//
//            fw.write("Transitions:")
//
//            for (i in 0..automaton.States.size - 1) {
//                val state = automaton.States[i]
//                for (j in 0..state.Transitions.size - 1) {
//                    val transition = state.Transitions[j]
//                    if (automaton.States.size - 1 == i && state.Transitions.size - 1 == j) {
//                        fw.write(state.Name + "," + transition.DestinyState.Name + "," + transition.Symbol)
//                    } else {
//                        fw.write(state.Name + "," + transition.DestinyState.Name + "," + transition.Symbol + ";")
//                    }
//
//                }
//            }
//
//            fw.close()
//        } catch (e: Exception) {
//            e.printStackTrace()
//        }
//
//        return false
//    }
}
